Record and use slice index for stacked netdcf + usability fixes

Kirill888 opened the pull request:
### Reason for this pull request

- Usability fixes for `datacube dataset add/update`
- Better handling of multi-tenanted files (i.e. stacked netcdf)

### Proposed changes

1. Remove product matching rules options from `datacube dataset update`, it operates on existing datasets only, it should need to auto-match anything.
2. Remove `--match-rules` option from `datacube dataset add`, this one wasn't being tested, nor was it documented. 
3. Deprecate `--auto-match` option, it was on by default anyway.
4. Rename `--dtype` to `--product` (discussed in #450)
5. `read_documents` can report url that includes part of the multi-tenant file form which documet was read
6. Indexing/update process records time slice index as part of the location URI, when working with stacked netcdf files
7. Stacker records time slice index as part of the location URI
8. When loading data use recorded index instead of searching for best matching timestamp



### Incidental changes

- Enable compression of metadata text during ingestion #452 
- Fix chunking configuration for strings in netcdf files  #453 

------------------------------------------------------------------------
 - [x] Closes #415 #447 #448 #452 #453 #454 #445
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

Kirill888 committed:
Unpin boto3 in setup.py

Kirill888 committed:
Print more info when plugin driver fails to resolve

I had some version pinning issues before and had to step through debugger to
understand them, with this it should be more obvious what is failing at the cost
of noisier output when things are bad.

Kirill888 committed:
Linting in cli app

- removed some unused imports
- whitespace between functions

Kirill888 committed:
Remove unused imports from stacker

Kirill888 committed:
Remove --match-rules option, deprecate --auto-match option

See issues #447 and #448

Kirill888 committed:
Add helper method to remove lineage sources from the document

Kirill888 committed:
Remove product matching options from dataset update cli command

Dataset update operates on existing datasets only, so no need to auto-match
anything, every dataset should be already in the database and associated with
some product.

Also dataset update won't load and resolve lineage sources anymore, internal
update apis don't really support adding new lineage data as part of an `update`
call anyway, although it is part of `can_update` check.

What this means is that there is a possibly change in behaviour when new version
of dataset has changes in the lineage document. Previously this would be
reported as incompatible change, now it won't be detected at all.

Kirill888 committed:
Stop using redundant --auto-match option in tests

Also add tests for `dataset update` and add "tests" for dry runs, not really
proper tests, as in they don't verify that dry-run is not changing database in
any way, but at least it runs that code.

Kirill888 committed:
Remove references to --auto-match from documentation

Kirill888 committed:
Formatting fixes, remove unused imports

Kirill888 committed:
Add utility function map_with_lookahead

This is useful when you need to process single element sequences differently
from multi-element sequences. Will be used in read_documents.

Kirill888 committed:
Refactor read_documents function

this is in preparation for better support for multi-document files

Kirill888 committed:
Add test for read_documents

Kirill888 committed:
Simplify map_with_lookahead

Kirill888 committed:
Correcting error in read_documents, netcdf case

Refactor introduced an error, raw strings were returned instead of parsed
documents when input is netcdf file. None of the tests picked them up, not even
integration tests. Adding tests for reading yaml docs from netcdf files.

Kirill888 committed:
Default to identity transform in map_with_lookahead

Kirill888 committed:
Add extra flag to read_documents to optionally return URIs

Instead of generating (Path, doc) tuples, generate (URI, doc) tuples, where URI
includes `#part=idx` for multi-tenant files (idx is 0 based)

Default behaviour remains the same.

Kirill888 committed:
Renamed --dtype,-t to --product,-p

Old options is still available, but will result in a warning message being
printed if used.

Kirill888 committed:
Test for dataset add option name remapping

Kirill888 committed:
Make sure path is absolute before converting to URI

Kirill888 committed:
Using `#part=idx` uris in dataset add/update for multi-part files

Basically using `uri=True` option when reading metadata documents, this will
return uris with part number populated when reading from multi-part files. This
means we can index stacked netcdfs and get different location for each time
slice. Reading code will just ignore them for now, but this will come next.

Kirill888 committed:
Fix chunking settings for string variables #453

Kirill888 committed:
Enable compression of metadata text in ingestor

part of #452

Kirill888 committed:
Support newer dask `dask.async` -> `dask.local`

See issue #454

Kirill888 committed:
Take uri building logic into a stand-alone function

Kirill888 committed:
Expand xr_apply function in backwards compatible way

- Default output dtype to be src dtype if not supplied
- Add option to use numeric index as well as labeled index in the callback

Kirill888 committed:
Stacker records index of each dataset as part of uri fragment

Kirill888 committed:
Remove stale commented out code

Kirill888 committed:
Remove unused imports in storage.py

Kirill888 committed:
Stop using compat module for url-related libs in storage.py

Now that we are py3 only start removing un-needed indirection.

Kirill888 committed:
Stop using compat module for url-related libs in utils lib

Kirill888 committed:
Use `#part=idx` to resolve time-slice when available

When reading from netcdf only go searching for a timestamp if slice index is not
recorded.

Kirill888 committed:
Enable dataset metadata compression in stacker #452

omad commented on datacube/utils/__init__.py:
```diff
@@ -238,7 +291,28 @@ def remove_implicit_resolver(cls, tag_to_remove):
+        return doc
+
+    if inplace:
+        doc['lineage']['source_datasets'] = {}
```
Unfortunately, this (and the rest) of dataset documents is configurable with the MetadataType thing. See https://github.com/opendatacube/datacube-core/blob/fbefaa5f95704160133cd45df4cd5a12e1ea88f9/datacube/index/default-metadata-types.yaml#L54 for where it is defined for the `eo` metadata type. (Lodge complaints with @jeremyh. :) I can see the appeal of configurability, but I think other metadata standards will be *more* different than what we can simply configure away, and we pay a significant cost in complexity...)

Fortunately, in the two places this function is used it does have access to the Product, which is of a particular MetadataType, so can lookup what offset `sources` will be at. Still a total pain though...

Kirill888 replied:
Good point, I keep forgetting about this indirection. It does mean that `sources` can not be skipped until top-level product is resolved (auto-matched), not a problem in `update`, but we do need this in `add` as well. I'll have a look what that implies.

omad reviewed (changes requested):
All looks amazing, except for the annoyingness with dataset document structure.

Can you please also copy the proposed and incidental changes into the `whats_new` doc.

Kirill888 committed:
Fix `without_lineage_sources` api/behaviour

Don't assume `eo` datatype when skipping sources, where lineage sources are
stored is defined by metadata type. So it's needed to erase them.

Kirill888 committed:
Update whats new document

Kirill888 committed:
Merge branch 'develop' into stacking-fixes

Kirill888 committed:
Merge develop into stacking-fixes

omad reviewed (approved):
Thank you! 👍
